Issue #872: DSWx-NI triggering logic

chrisjrd opened the pull request:
## Purpose
implement DSWx-NI grace period boilerplate.
## Proposed Changes
- [CHANGE] DSWx-NI query and download to mark records as submitted_for_download to support future grace period implementation.

## Issues
#872 
## Testing
tested locally

chrisjrd committed:
feat(DSWx-NI): DSWx-NI workflow query and download steps

chrisjrd committed:
fix(DSWx-NI): gcov download reference to sciflo job queue. hard-coded.

chrisjrd committed:
fix(DSWx-NI): fix catalogue-ing to prevent duplicates, and support "full" mode for local testing.

chrisjrd committed:
refactor(DSWx-NI): minor refactor

chrisjrd committed:
refactor(DSWx-NI): moved gcov file to gcov package

chrisjrd committed:
fix(DSWx-NI): gcov download job retrieving extra records

chrisjrd committed:
fix(DSWx-NI): marking submitted downloads for easier grace period implementation

chrisjrd committed:
fix(DSWx-NI): refactor code to support grace period

chrisjrd committed:
Merge branch 'refs/heads/develop' into issue_872

# Conflicts:
#	data_subscriber/parser.py

hhlee445 commented:
Getting an error  with a command 

`python3 ~/mozart/ops/opera-pcm/data_subscriber/daac_data_subscriber.py query -c NISAR_L2_GCOV_BETA_V1  --native-id=NISAR_L2_PR_GCOV_088_039_D_085_2005_SHSH_A_20251114T220420_20251114T220427_T00407_N_P_J_001 --endpoint=UAT --job-queue=opera-job_worker-gcov_download `


```
[2026-01-07 19:23:07,843: INFO/supply_job_id] Running outside of job context. Generating random job ID. job_id='94694eb7-1d57-456b-bde3-046ebe6c3cad'
Traceback (most recent call last):
  File "/export/home/hysdsops/mozart/ops/opera-pcm/data_subscriber/daac_data_subscriber.py", line 165, in <module>
    main()
  File "/export/home/hysdsops/mozart/ops/opera-pcm/util/exec_util.py", line 35, in wrapper
    status = func(*args, **kwargs)
  File "/export/home/hysdsops/mozart/ops/opera-pcm/data_subscriber/daac_data_subscriber.py", line 45, in main
    run(sys.argv)
  File "/export/home/hysdsops/mozart/ops/opera-pcm/data_subscriber/daac_data_subscriber.py", line 73, in run
    results["query"] = run_query(args, token, es_conn, cmr, job_id, settings)
  File "/export/home/hysdsops/mozart/ops/opera-pcm/data_subscriber/daac_data_subscriber.py", line 104, in run_query
    cmr_query = NisarGcovCmrQuery(args, token, es_conn, cmr, job_id, settings)
  File "/export/home/hysdsops/mozart/ops/opera-pcm/data_subscriber/gcov/gcov_query.py", line 17, in __init__
    super().__init__(args, token, es_conn, cmr, job_id, settings)
  File "/export/home/hysdsops/mozart/ops/opera-pcm/data_subscriber/query.py", line 41, in __init__
    self.query_replacement_file = args.query_replacement_file
AttributeError: 'Namespace' object has no attribute 'query_replacement_file'
```

chrisjrd committed:
fix(DSWx-NI): re-add param that was dropped during the last commit/merge

hhlee445 commented:
DSWX-NI SCIFLO jobs failed with 

```
Exception Type: <class 'ValueError'>
Exception Value: Number of processes must be at least 1
Traceback (most recent call last):
  File "/root/verdi/ops/sciflo-1.3.7/sciflo/grid/workUnit.py", line 130, in run
    result = self._run()
  File "/root/verdi/ops/sciflo-1.3.7/sciflo/grid/workUnit.py", line 172, in _run
    return func(*funcArgs)
  File "/tmp/scifloWork-ops/sciflowuid-15818421-01538571-0788082026-ae0179ad4569b9b623d8273d60e8d13c/run_pge_docker.py", line 87, in submit_pge_job
    job_json = cls_object.submit_job()
  File "/root/verdi/ops/chimera-2.2.4/chimera/pge_job_submitter.py", line 216, in submit_job
    job_json = self.perform_adaptation_tasks(job_json)
  File "/home/ops/verdi/ops/opera-pcm/opera_chimera/opera_pge_job_submitter.py", line 99, in perform_adaptation_tasks
    with get_context("spawn").Pool(
  File "/opt/conda/lib/python3.9/multiprocessing/context.py", line 119, in Pool
    return Pool(processes, initializer, initargs, maxtasksperchild,
  File "/opt/conda/lib/python3.9/multiprocessing/pool.py", line 205, in __init__
    raise ValueError("Number of processes must be at least 1")
ValueError: Number of processes must be at least 1
```


It submitted jobs over and over again. Need to check the trigger logic

Commit:
Merge remote-tracking branch 'origin/develop' into issue_872

Commit:
patch issue with parallel localization

localize_urls=[] would lead to error

RKuttruff commented:
@hhlee445 [This](https://github.com/nasa/opera-sds-pcm/pull/1331#issuecomment-3725999792) isn't a triggering issue. This is a bug that was introduced with the parallelized localization work I did (essentially trying to create a process pool with 0 processes). I've pushed a fix

Commit:
fix download job matching

Commit:
force es index refresh after cataloging

Commit:
add dswx-ni triggering options to settings

Commit:
delete duplicate dswx-ni queue in tf

Commit:
dont't give all batch ids to each triggered dl job

Commit:
caching for mgrs db frame lookup

may not be needed

Commit:
First attempt at triggering login implementation

Commit:
Catalog CMR results before dl determination

I worry there may be side effects here... but will commit this for testing

Commit:
Add some logging

Commit:
tweak batch id maps to restrict to submitted batches

Commit:
filtering by land_ocean_flag == "water"

Commit:
filtering by land_ocean_flag == "water"

store lof in doc in catalog step

Commit:
fix es queries

Commit:
try fix batch_id maps for marking coverage

Commit:
add coverage to download job sub mark

Commit:
try mark downloaded in dl job

Commit:
fix mistaken computation

Commit:
fix deduping for download jobs

otherwise download job for sets that were previously partially triggered will be deduped in a way that will damage the GRQ index

Commit:
some cleanup

also reduce some debug log messages from info to debug

Commit:
Pass all matching docs & granules down

Commit:
Pass all matching docs & granules down

Commit:
Pass all matching docs & granules down

Commit:
add https url field(s) to gcov

Commit:
attempt at https localization

Commit:
rework

Commit:
some logging

Commit:
fix failing unit test cases

Commit:
some additional fixes to failing unit test cases

Commit:
add some log lines

should hopefully clarify situations where granules are catalogued but no evaluation happens due to all granules being over water

Commit:
more logging for evaluation logic

hhlee445 committed:
Add window_delta in parser.py

hhlee445 committed:
Add window_delta in parser.py

hhlee445 commented:
The following command:

```
python3 ~/mozart/ops/opera-pcm/data_subscriber/daac_data_subscriber.py query \
  -c NISAR_L2_GCOV_BETA_V1 \
  --job-queue=opera-job_worker-gcov_query \
  --chunk-size 1 \
  --native-id=NISAR_L2_PR_GCOV_015_156_A_010_2005_DVDV_A_20230619T000803_20230619T000818_T00408_N_P_J_001 \
  --endpoint=UAT
```

is expected to submit one DSWx-NI job for MS_156_10$15.

Frames 10 and 11 are available in CMR:

- NISAR_L2_PR_GCOV_015_156_A_010_2005_DVDV_A_20230619T000803_20230619T000818_T00408_N_P_J_001
- NISAR_L2_PR_GCOV_015_156_A_011_2005_DVDV_A_20230619T000816_20230619T000835_T00408_N_P_J_001

CMR query reference:
https://cmr.uat.earthdata.nasa.gov/search/granules.umm_json?provider=ASF&ShortName=NISAR_L2_GCOV_BETA_V1&native-id=NISAR_L2_PR_GCOV_015_156_A_*00408_N_P_J_001&options[native-id]

We need to query all available frames from CMR and pass the corresponding S3 paths into the DSWx-NI SCIFLO job.

hhlee445 committed:
Merge branch 'develop' into issue_872

RKuttruff committed:
bperp docker build plain progress

RKuttruff committed:
Merge remote-tracking branch 'origin/develop' into issue_872

RKuttruff committed:
update gcov utils

accept literal strings as input

RKuttruff committed:
add new gcov utils

orbit direction and acquisition time range

RKuttruff committed:
add track frame db func to map (track, frame) -> frames

RKuttruff committed:
update handling of gcov native id query

RKuttruff committed:
properly format native id fields

RKuttruff commented:
@hhlee445 I've implemented a fix for the native ID issue. The command you provided now submits the expected job

<details>
<summary>Log output</summary>

```
$ python3 ~/mozart/ops/opera-pcm/data_subscriber/daac_data_subscriber.py query   -c NISAR_L2_GCOV_BETA_V1   --job-queue=opera-job_worker-gcov_query   --chunk-size 1   --native-id=NISAR_L2_PR_GCOV_015_156_A_010_2005_DVDV_A_20230619T000803_20230619T000818_T00408_N_P_J_001   --endpoint=UAT
[2026-02-27 22:21:01,633: INFO/get_logger] Initial logging configuration complete
[2026-02-27 22:21:01,633: INFO/get_logger] Log level set to INFO
[2026-02-27 22:21:01,879: INFO/configure_library_loggers] Added logging filter for hysds_commons
[2026-02-27 22:21:01,880: INFO/configure_library_loggers] Added logging filter for elasticsearch
[2026-02-27 22:21:01,880: INFO/configure_library_loggers] Added logging filter for opensearch
[2026-02-27 22:21:01,880: INFO/configure_library_loggers] Configuring boto3 logger
[2026-02-27 22:21:01,880: INFO/configure_library_loggers] Filtering (ignore) ElasticsearchWarning
[2026-02-27 22:21:01,880: INFO/configure_library_loggers] Filtering (ignore) CryptographyDeprecationWarning
[2026-02-27 22:21:01,880: INFO/get_es_connection] aws_es=False
[2026-02-27 22:21:01,880: INFO/get_es_connection] es_url='http://100.104.40.5:9200'
[2026-02-27 22:21:01,880: INFO/get_es_connection] es_engine='opensearch'
[2026-02-27 22:21:01,882: INFO/log_request_success] HEAD http://100.104.40.5:9200/ [status:200 request:0.002s]
[2026-02-27 22:21:01,882: INFO/supply_job_id] Running outside of job context. Generating random job ID. job_id='f0619821-79b6-4d2c-b544-8eafe6d992e1'
[2026-02-27 22:21:02,672: INFO/_delete_token] CMR token successfully deleted
[2026-02-27 22:21:02,797: INFO/load_mgrs_track_frame_db] Loading MGRS track frame database from ParseResult(scheme='s3', netloc='opera-ancillaries', path='/dswx_ni/MGRS_collection_db_DSWx-NI_v0.1.sqlite', params='', query='', fragment='')
[2026-02-27 22:21:03,207: INFO/_get_query_func] Selected data source: CMR
[2026-02-27 22:21:03,207: INFO/query_cmr] Query CMR for NISAR L2 GCOV products with timerange: DateTimeRange(start_date='1900-01-01T00:00:00Z', end_date='2026-02-27T22:21:03Z')
[2026-02-27 22:21:03,257: INFO/load_mgrs_track_frame_db] Loading MGRS track frame database from ParseResult(scheme='s3', netloc='opera-ancillaries', path='/dswx_ni/MGRS_collection_db_DSWx-NI_v0.1.sqlite', params='', query='', fragment='')
[2026-02-27 22:21:03,439: INFO/async_query_cmr] track_number=156, cycle_number=15, orbit_direction='A'
[2026-02-27 22:21:03,448: INFO/async_query_cmr] Matched native ID to frames [9, 10, 11, 12]
[2026-02-27 22:21:03,449: INFO/async_query_cmr] Determined time range 2023-06-19 00:08:03 - 2023-06-19 00:08:18
[2026-02-27 22:21:03,449: INFO/async_query_cmr] temporal_range='2023-06-18T23:08:03Z,2023-06-19T01:08:18Z'
[2026-02-27 22:21:03,449: INFO/async_query_cmr] Querying CMR. endpoint: cmr.uat.earthdata.nasa.gov  provider: None
[2026-02-27 22:21:03,450: INFO/async_cmr_post] Issuing request. This may take a while depending on search page size and number of pages/results.
[2026-02-27 22:21:04,052: INFO/async_cmr_post] Reached end of CMR search results. Ending query.
[2026-02-27 22:21:04,053: INFO/async_query_cmr] CMR Query Complete. Found 5 granule(s)
[2026-02-27 22:21:04,053: INFO/async_query_cmr] QUERY RESULTS 1 to 5 of 5:
[2026-02-27 22:21:04,053: INFO/async_query_cmr] ('NISAR_L2_PR_GCOV_015_156_A_011_2005_DVDV_A_20230619T000817_20230619T000835_T00406_M_P_J_001', 'revision 17')
[2026-02-27 22:21:04,053: INFO/async_query_cmr] ('NISAR_L2_PR_GCOV_015_156_A_011_2005_DVDV_A_20230619T000817_20230619T000835_T05000_N_P_J_001', 'revision 1')
[2026-02-27 22:21:04,053: INFO/async_query_cmr] ('NISAR_L2_PR_GCOV_015_156_A_011_2005_DVDV_A_20230619T000816_20230619T000835_T00408_N_P_J_001', 'revision 1')
[2026-02-27 22:21:04,053: INFO/async_query_cmr] ('NISAR_L2_PR_GCOV_015_156_A_010_2005_DVDV_A_20230619T000803_20230619T000818_T00408_N_P_J_001', 'revision 1')
[2026-02-27 22:21:04,053: INFO/async_query_cmr] ('NISAR_L2_PR_GCOV_015_156_A_010_2005_DVDV_A_20230619T000803_20230619T000818_T05000_N_P_J_001', 'revision 13')
[2026-02-27 22:21:04,054: INFO/run_query] Granule Cataloguing STARTED
[2026-02-27 22:21:04,129: INFO/_catalog_granules] Cataloging GCOV granule: NISAR_L2_PR_GCOV_015_156_A_011_2005_DVDV_A_20230619T000817_20230619T000835_T00406_M_P_J_001
[2026-02-27 22:21:04,376: INFO/_catalog_granules] Cataloging GCOV granule: NISAR_L2_PR_GCOV_015_156_A_011_2005_DVDV_A_20230619T000817_20230619T000835_T00406_M_P_J_001
[2026-02-27 22:21:04,386: INFO/_catalog_granules] Cataloging GCOV granule: NISAR_L2_PR_GCOV_015_156_A_011_2005_DVDV_A_20230619T000817_20230619T000835_T00406_M_P_J_001
[2026-02-27 22:21:04,395: INFO/_catalog_granules] Cataloging GCOV granule: NISAR_L2_PR_GCOV_015_156_A_011_2005_DVDV_A_20230619T000817_20230619T000835_T05000_N_P_J_001
[2026-02-27 22:21:04,404: INFO/_catalog_granules] Cataloging GCOV granule: NISAR_L2_PR_GCOV_015_156_A_011_2005_DVDV_A_20230619T000817_20230619T000835_T05000_N_P_J_001
[2026-02-27 22:21:04,412: INFO/_catalog_granules] Cataloging GCOV granule: NISAR_L2_PR_GCOV_015_156_A_011_2005_DVDV_A_20230619T000817_20230619T000835_T05000_N_P_J_001
[2026-02-27 22:21:04,422: INFO/_catalog_granules] Cataloging GCOV granule: NISAR_L2_PR_GCOV_015_156_A_011_2005_DVDV_A_20230619T000816_20230619T000835_T00408_N_P_J_001
[2026-02-27 22:21:04,431: INFO/_catalog_granules] Cataloging GCOV granule: NISAR_L2_PR_GCOV_015_156_A_011_2005_DVDV_A_20230619T000816_20230619T000835_T00408_N_P_J_001
[2026-02-27 22:21:04,439: INFO/_catalog_granules] Cataloging GCOV granule: NISAR_L2_PR_GCOV_015_156_A_011_2005_DVDV_A_20230619T000816_20230619T000835_T00408_N_P_J_001
[2026-02-27 22:21:04,447: INFO/_catalog_granules] Cataloging GCOV granule: NISAR_L2_PR_GCOV_015_156_A_010_2005_DVDV_A_20230619T000803_20230619T000818_T00408_N_P_J_001
[2026-02-27 22:21:04,456: INFO/_catalog_granules] Cataloging GCOV granule: NISAR_L2_PR_GCOV_015_156_A_010_2005_DVDV_A_20230619T000803_20230619T000818_T00408_N_P_J_001
[2026-02-27 22:21:04,464: INFO/_catalog_granules] Cataloging GCOV granule: NISAR_L2_PR_GCOV_015_156_A_010_2005_DVDV_A_20230619T000803_20230619T000818_T00408_N_P_J_001
[2026-02-27 22:21:04,474: INFO/_catalog_granules] Cataloging GCOV granule: NISAR_L2_PR_GCOV_015_156_A_010_2005_DVDV_A_20230619T000803_20230619T000818_T05000_N_P_J_001
[2026-02-27 22:21:04,482: INFO/_catalog_granules] Cataloging GCOV granule: NISAR_L2_PR_GCOV_015_156_A_010_2005_DVDV_A_20230619T000803_20230619T000818_T05000_N_P_J_001
[2026-02-27 22:21:04,489: INFO/_catalog_granules] Cataloging GCOV granule: NISAR_L2_PR_GCOV_015_156_A_010_2005_DVDV_A_20230619T000803_20230619T000818_T05000_N_P_J_001
[2026-02-27 22:21:04,497: INFO/_catalog_granules] Cataloged: 15 GCOV granules
[2026-02-27 22:21:04,497: INFO/_catalog_granules] Performing index refresh
[2026-02-27 22:21:04,517: INFO/log_request_success] POST http://100.104.40.5:9200/nisar_gcov_catalog*/_refresh [status:200 request:0.020s]
[2026-02-27 22:21:04,517: INFO/_catalog_granules] Performed index refresh
[2026-02-27 22:21:04,517: INFO/_catalog_granules] Getting all unsubmitted docs not over water
[2026-02-27 22:21:04,523: INFO/_catalog_granules] Found len(unsubmitted_docs)=12
[2026-02-27 22:21:04,606: INFO/run_query] Granule Cataloguing FINISHED
[2026-02-27 22:21:04,641: INFO/determine_download_granules] Found 15 GCOV granules
[2026-02-27 22:21:04,641: INFO/determine_download_granules] Found 4 unique MGRS sets and cycle numbers
[2026-02-27 22:21:04,641: INFO/determine_download_granules] Getting all unsubmitted docs not over water
[2026-02-27 22:21:04,646: INFO/determine_download_granules] Found len(unsubmitted_docs)=12
[2026-02-27 22:21:04,646: INFO/determine_download_granules] Getting all partial docs not over water
[2026-02-27 22:21:04,649: INFO/determine_download_granules] Found len(incomplete_docs)=0
[2026-02-27 22:21:04,649: INFO/determine_download_granules] Grouped unsubmitted and incomplete docs into 3 MGRS sets
[2026-02-27 22:21:04,649: INFO/determine_download_granules] 0 MGRS sets will be removed from consideration since they contain no new frames
[2026-02-27 22:21:04,649: INFO/_evaluate_mgrs_set_id_cycle_indices] Evaluating triggers for 3 MGRS tile sets
[2026-02-27 22:21:04,649: INFO/_evaluate_mgrs_set_id_cycle_indices] Triggering params: min_num_frames=1 coverage_target=None grace_mins=210
[2026-02-27 22:21:04,649: INFO/_evaluate_mgrs_set_id_cycle_indices] Evaluating mgrs_set_id_cycle_index='MS_156_10$15'
[2026-02-27 22:21:04,657: INFO/_evaluate_mgrs_set_id_cycle_indices] Expected frames for MGRS set: {10, 11}. Available frames: {10, 11}
[2026-02-27 22:21:04,657: INFO/_evaluate_mgrs_set_id_cycle_indices] Triggering MS_156_10$15 as it is fully covered
[2026-02-27 22:21:04,657: INFO/_evaluate_mgrs_set_id_cycle_indices] Evaluating mgrs_set_id_cycle_index='MS_156_11$15'
[2026-02-27 22:21:04,664: INFO/_evaluate_mgrs_set_id_cycle_indices] Expected frames for MGRS set: {10, 11, 12}. Available frames: {10, 11}
[2026-02-27 22:21:04,664: INFO/_evaluate_mgrs_set_id_cycle_indices] Frame set MS_156_11$15 has sufficient coverage to trigger
[2026-02-27 22:21:04,665: INFO/_evaluate_mgrs_set_id_cycle_indices] Frame set still within grace period (grace_period_minutes_remaining=datetime.timedelta(seconds=12599, microseconds=824972)) mgrs_set_id_cycle_index='MS_156_11$15'
[2026-02-27 22:21:04,665: INFO/_evaluate_mgrs_set_id_cycle_indices] Evaluating mgrs_set_id_cycle_index='MS_156_9$15'
[2026-02-27 22:21:04,672: INFO/_evaluate_mgrs_set_id_cycle_indices] Expected frames for MGRS set: {9, 10}. Available frames: {10}
[2026-02-27 22:21:04,673: INFO/_evaluate_mgrs_set_id_cycle_indices] Frame set MS_156_9$15 has sufficient coverage to trigger
[2026-02-27 22:21:04,673: INFO/_evaluate_mgrs_set_id_cycle_indices] Frame set still within grace period (grace_period_minutes_remaining=datetime.timedelta(seconds=12599, microseconds=800954)) mgrs_set_id_cycle_index='MS_156_9$15'
[2026-02-27 22:21:04,754: INFO/submit_gcov_download_job_submission_handler] Triggering GCOV jobs for 1 unique MGRS sets and cycle numbers to process
[2026-02-27 22:21:04,859: INFO/mark_products_as_download_job_submitted] Marking {'MS_156_10$15'} products as download job-submitted, in bulk
[2026-02-27 22:21:04,933: INFO/log_request_success] POST http://100.104.40.5:9200/_bulk [status:200 request:0.074s]
[2026-02-27 22:21:04,966: INFO/log_request_success] POST http://100.104.40.5:9200/nisar_gcov_catalog*/_refresh [status:200 request:0.033s]
[2026-02-27 22:21:04,967: INFO/run_query] 1 download jobs succeeded=['d8865c7a-d8f3-4854-be0d-0b14eca6413d']
[2026-02-27 22:21:04,967: INFO/run_query] 0 download jobs failed=[]
[2026-02-27 22:21:04,967: INFO/run] len(results)=1
[2026-02-27 22:21:04,967: INFO/run] END
```

</details>

RKuttruff committed:
tweak log line

hhlee445 committed:
Merge branch 'develop' into issue_872
